Add ROUTE_CAVEAT_CONFIGS as a pre-built configuration option

aaroncj1 opened the pull request:
Added a new pre-built config to restrict access to specific routes. There is an option to add a master route to your server that when you pay that invoice will create an LSAT that gives you access to all routes. To do this, add MASTER_ROUTE to your environment variables and set it equal to your desired master route.

Commit:
added route configuration

Commit:
Add ROUTE_CAVEAT_CONFIGS documentation to README

Commit:
Add ROUTE_CAVEAT_CONFIGS documentation to README

bucko13 commented on package.json:
```diff
@@ -17,7 +17,7 @@
     "clean": "rm -rf dist && mkdir dist && rm -rf bin && mkdir bin",
     "lint": "tsc --noEmit --project tsconfig.json && eslint --ext .ts src --quiet",
-    "test": "LOG_LEVEL=none nyc ts-mocha -p tsconfig.json tests/**/*.spec.ts",
+    "test": "LOG_LEVEL=none MASTER_ROUTE=/master nyc ts-mocha -p tsconfig.json tests/**/*.spec.ts",
```
I think ideally the tests would be able to stay agnostic about "runtime" details like this. So if this was something we needed to set for specific tests it should stay scoped to those particular tests.

aaroncj1 replied:
Good point, thanks I pushed change

bucko13 commented on src/satisfiers/index.ts:
```diff
@@ -1,2 +1,3 @@
 export { default as originSatisfier } from './origin'
 export { challengeSatisfier } from './challenge'
+export { default as routeSatisfier } from './route'
```
I would just export the const `ROUTE_CAVEAT_CONFIGS` as is without default from its file and then this can be consistent with the others

```suggestion
export { routeSatisfier } from './route'
```

bucko13 commented on src/satisfiers/route.ts:
```diff
@@ -0,0 +1,19 @@
+  },
+}
+
+export default routeSatisfier
```
same here, prefer named exports (would probably be better if there was an eslint rule for this. sorry)

bucko13 commented on yarn.lock:
```diff
@@ -3,16 +3,16 @@
-  resolved "https://registry.yarnpkg.com/@babel/code-frame/-/code-frame-7.8.3.tgz#33e25903d7481181534e12ec0a25f16b6fcf419e"
-  integrity sha512-a9gxpmdXtZEInkCSHUJDLHZVBgb1QS0jhss4cPP93EW7s+uC5bikET2twEF3KV+7rDblJcmNvTR7VJejqd2C2g==
+  "integrity" "sha512-a9gxpmdXtZEInkCSHUJDLHZVBgb1QS0jhss4cPP93EW7s+uC5bikET2twEF3KV+7rDblJcmNvTR7VJejqd2C2g=="
+  "resolved" "https://registry.npmjs.org/@babel/code-frame/-/code-frame-7.8.3.tgz"
```
As you didn't do any changes to package.json, there shouldn't be any changes to the yarn.lock. It looks like you might've done an `npm i` maybe? Either way this shouldn't be in the diff for this PR.

bucko13 commented on src/satisfiers/route.ts:
```diff
@@ -0,0 +1,19 @@
+  condition: 'route',
+  satisfyFinal: (caveat: Caveat, req: Request) => {
+    const path = req.path
+    if (caveat.condition === 'route' && (caveat.value === path || caveat.value === MASTER_ROUTE)) return true
```
I'm thinking that an alternative way to do the MASTER_ROUTE such that you don't have to add global values to the process environment, and should simplify testing such that you don't have to add an environment variable to get tests to work, is to extend the `BoltwallConfig` (which gets attached to the request object so that other middlewares can have access to it, see [this](https://stackoverflow.com/questions/37377731/extend-express-request-object-using-typescript) SO question for ideas about extending the `Request` object).

This makes testing much easier because rather than adding global variables you can just mock different request objects/boltwall configs and that way also test different values for this configuration.

aaroncj1 replied:
Awesome. This is a much better way, thank you. I didn't see that boltwallConfig was getting added to the request. I pushed changes to fix this. Thanks!

bucko13 commented on tests/configs.spec.ts:
```diff
@@ -281,4 +281,103 @@ describe('configs', () => {
+    })
+
+    it('should create a caveat that restricts access by route and be able to satisfy it', () => {
+      const request: any = { path: '/path' }
```
not a fan of `any` if it can be avoided. I think it should be ok leaving this not typed and the type will just be inferred from the object which should be fine for the test

bucko13 reviewed (changes requested):
Looks good! Thanks for putting this together. Mostly just a couple nits. The only structural change that I think would be good is to move away from using the environment variable for the MASTER_ROUTE. My suggestion is to put it in the request object/boltwallConfig, but open to others. This will clean up the tests a bit too since you can make it more explicit. 

Question: how do you think this should handle subroutes or generally more complex routing situations? The way I'm reading this, if you went to route `/protected` and paid the LSAT, but then wanted to access `GET /protected/images` that wouldn't work. Maybe that's desirable, but maybe we want to allow for any paths within a particular "[router](https://expressjs.com/en/guide/routing.html#express-router)" (to use express.js terminology).

Commit:
moved masterRoute to boltwall config and add option for subroutes

aaroncj1 commented:
> Looks good! Thanks for putting this together. Mostly just a couple nits. The only structural change that I think would be good is to move away from using the environment variable for the MASTER_ROUTE. My suggestion is to put it in the request object/boltwallConfig, but open to others. This will clean up the tests a bit too since you can make it more explicit.
> 
> Question: how do you think this should handle subroutes or generally more complex routing situations? The way I'm reading this, if you went to route `/protected` and paid the LSAT, but then wanted to access `GET /protected/images` that wouldn't work. Maybe that's desirable, but maybe we want to allow for any paths within a particular "[router](https://expressjs.com/en/guide/routing.html#express-router)" (to use express.js terminology).

This is great. Thank you for the review. I moved the master route configuration to BoltwallConfig and added a allowSubroutes boolean to turn on the ability to allow LSATs to be used on subroutes of the authorized route.

bucko13 reviewed (changes requested):
Mostly just a couple of small nits left. Looking good though!

aaroncj1 commented on src/satisfiers/route.ts:
```diff
@@ -0,0 +1,22 @@
+    const master = req.boltwallConfig?.masterRoute
+    let pathMatches
+    if(req.boltwallConfig?.allowSubroutes) {
+      pathMatches = (req.path+'/').startsWith(caveat.value+'/')
```
I added the `+'/'` at the end to make sure that an LSAT to access `/route` wouldn't give you access to other routes that started the same, like `/routeTwo`.

Splitting the route and comparing first item would be cleaner for 1 subroute deep. But what about longer subroute paths? For example, an LSAT with access to `/images/dogs` could access all the `/images` subroutes. 

I think the way it is written now gives a fuller validation, to be a valid subroute you need to start with exactly the original LSAT path. Access to `/images/dogs` provides access to `/images/dogs` and any route that starts with `/images/dogs/`

I'd love to hear your thoughts.

bucko13 replied:
Ah ha. I see what you mean. I guess the only other concern I'd have is if the config is somehow set with a path value that already has a trailing slash. Maybe they should be trimmed if they exist before getting set on the macaroon?

aaroncj1 commented on src/satisfiers/route.ts:
```diff
@@ -0,0 +1,22 @@
+    if(req.boltwallConfig?.allowSubroutes) {
+      pathMatches = (req.path+'/').startsWith(caveat.value+'/')
+    } else pathMatches = (caveat.value === req.path)
+    if (caveat.condition === 'route' && (pathMatches || caveat.value === master)) return true
```
Yes! I thought this was a interesting idea. Could be a useful way to provide all access pass to your endpoints.

I've also thought about how it would be cool to be able to combine a few of these configs together. Maybe you want both time caveat and route caveat. I'd have to think about it a bit more how to impliment. Maybe that could be another PR in the future.

bucko13 replied:
Yeah totally. The config object itself does allow for composability which you can see in the [type definition](https://github.com/Tierion/boltwall/blob/master/src/typings/configs.d.ts#L31-L39) where `getCaveats` and `caveatSatisfiers` can be arrays in which case each one would be added in sequence. So it would be nice to be able to pass in an array of configs you want to compose. There's some trickiness around the other config values which is probably why I didn't add that originally (e.g. invoice description, amounts, rate) but that could just be overridden as necessary.

Commit:
fixed typos and descriptions

bucko13 commented:
@aaroncj1 would you mind just pushing an empty commit to try and bump the travis status checks? Unfortunately it looks like there were some changes in travis that had the service down. I think everything is back up but it doesn't seem like it's possible to retrigger a build either in GH or travis (which is pretty frustrating).

Commit:
add check on request path

aaroncj1 commented:
> @aaroncj1 would you mind just pushing an empty commit to try and bump the travis status checks? Unfortunately it looks like there were some changes in travis that had the service down. I think everything is back up but it doesn't seem like it's possible to retrigger a build either in GH or travis (which is pretty frustrating).

Pushed a commit with a check on the ending `/` on the request path

Commit:
removed console log

bucko13 commented:
Good stuff! Thanks for the contribution.
